### PR TITLE
[subset] use charstrings directly from accelerator cache if mutability isn't needed.

### PR DESCRIPTION
**Pull request by garretrieger**
~18% improvement for large subset sizes.

BM_subset/subset_glyphs/SourceHanSans-Regular_subset.otf/10000_median                    -0.1794         -0.1796            54            44            54            44
BM_subset/subset_codepoints/SourceHanSans-Regular_subset.otf/10000_median                -0.1807         -0.1809            52            42            52            42

### CODE REVIEW

File: src/hb-subset-cff-common.hh [hunk 1]
```diff
@@ -1038,12 +1049,25 @@ struct subr_subsetter_t
 
   }
 
+  parsed_cs_str_t& get_parsed_charstring (unsigned i)
+  {
+    if (cached_charstrings.length) return *(cached_charstrings[i]);
```
Comment (behdad) [inline src/hb-subset-cff-common.hh:1054]: vector has operator bool so you can just say `if (cached_charstrings)`.

**Comment by behdad** [comment]
You are using a vector of pointers. I was thinking of not even using. Is that too much effort, to store a pointer to the vector-of-charstrings in the accel?

**Comment by behdad** [comment]
Shouldn't really matter for performance.

**Comment by garretrieger** [comment]
> You are using a vector of pointers. I was thinking of not even using. Is that too much effort, to store a pointer to the vector-of-charstrings in the accel?

The parsed_charstrings vector uses different indexing (the new gids) vs the cached one (old gids)

File: src/hb-subset-cff-common.hh (hunk 1 above)
Comment (garretrieger) [reply src/hb-subset-cff-common.hh]: Fixed

**Comment by behdad** [comment]
> > You are using a vector of pointers. I was thinking of not even using. Is that too much effort, to store a pointer to the vector-of-charstrings in the accel?
> 
> The parsed_charstrings vector uses different indexing (the new gids) vs the cached one (old gids)

I see. Sounds good.

**Pull request merged**